### PR TITLE
Aerogear 2333 rx refactor [WIP]

### PR DESCRIPTION
**Pull request by secondsun**
## Motivation

Right now we use a combination of RxStyle APIS (HttpRequest and Response) and Callbacks (Callback).  This PR seeks to unify those under a more thought out reactive style API.

JIRA: https://issues.jboss.org/browse/AGDROID-2333

## Description

This PR contains a Reactive API that is used for attaching multiple handlers to a request that emits a single calculated value when they are attached.  This is similar to a Single Observable from RxJava, but it is not the same.

## Progress

- [x] Request and Responder Interfaces
- [x] Synchronous Implementation
- [x] Asynchronous and Threading configurable Implementation
- [x] Ability to Cancel Requests
- [x] Ability to Cache Requests
- [x] Ability to Detach a Responder from a request
- [x] Unit Tests for all of the above
- [ ] Replace HttpRequest and HttpResponse
- [ ] Replace Callbacks

### CODE REVIEW

File: core/src/main/java/org/aerogear/mobile/core/reactive/CacheRequest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    private T cachedResult;
+    private Exception cachedException;
+
+    private List<AtomicReference<Responder<T>>> awaitingResponders =
+                    Collections.synchronizedList(new ArrayList<>());
+
+    public CacheRequest(Request<T> delegateTo) {
+        nonNull(delegateTo, "delegateTo");
+        this.delegateTo = delegateTo;
```
Comment (ziccardi) [inline core/src/main/java/org/aerogear/mobile/core/reactive/CacheRequest.java]: this can be changed to `this.delegateTo = nonNull(delegateTo, "delegateTo")`

File: core/src/main/java/org/aerogear/mobile/core/reactive/CallableRequest.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ */
+public final class CallableRequest<T> extends AbstractRequest<T> {
+
+    private static final String ERROR_TAG = "UNCAUGHT_EXCEPTION";
+    private final Callable<T> callable;
+    private Thread callableThread = null;
+
+    public CallableRequest(Callable<T> callable) {
+        SanityCheck.nonNull(callable, "callable");
+        this.callable = callable;
```
Comment (ziccardi) [inline core/src/main/java/org/aerogear/mobile/core/reactive/CallableRequest.java]: this can be `this.callable = nonNull(callable, "callable");`

File: core/src/main/java/org/aerogear/mobile/core/reactive/CallableRequest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        Responder<T> responder = responderRef.get();
+        if (responder == null) { // responder may have been disconnected while the calculation was
+                                 // performed.
+            return this;
+        }
+
+        try {
+            if (exception == null) {
+                responder.onResult(value);
```
Comment (ziccardi) [inline core/src/main/java/org/aerogear/mobile/core/reactive/CallableRequest.java]: if `responder` have been disconnected after the previous check, but before this call we will get a `NullPointerException`
I think we should synchronize this block

Comment (secondsun) [reply core/src/main/java/org/aerogear/mobile/core/reactive/CallableRequest.java]: The local repsonder reference can't be set to null by an external thread.  The AtomicReference may lose its reference, but we don't care about it anymore.

File: core/src/main/java/org/aerogear/mobile/core/reactive/CallableRequest.java (hunk 2 above)
Comment (secondsun) [reply core/src/main/java/org/aerogear/mobile/core/reactive/CallableRequest.java]: I smell a new lint rule we can implement.

File: core/src/main/java/org/aerogear/mobile/core/reactive/CacheRequest.java (hunk 1 above)
Comment (secondsun) [reply core/src/main/java/org/aerogear/mobile/core/reactive/CacheRequest.java]: See https://github.com/aerogear/aerogear-android-sdk/pull/160

Comment (secondsun) [reply core/src/main/java/org/aerogear/mobile/core/reactive/CacheRequest.java]: image

File: core/build.gradle [hunk 4]
```diff
@@ -47,6 +47,8 @@ dependencies {
     implementation "com.android.support:appcompat-v7"
     implementation "com.squareup.okhttp3:okhttp"
 
+    lintChecks files('libs/checks.jar')
```
Comment (danielpassos) [inline core/build.gradle]: Is this related to this PR of with the lint one?

File: core/src/main/java/org/aerogear/mobile/core/Request.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+     * @param responder a responder
+     * @return a chainable instance of Request, not guaranteed to be `this`
+     */
+    Request<T> respondWith(Responder<T> responder);
+
+    /**
+     * Requests can be run off the calling thread. This method configures that.
+     *
+     * @param executorService service to Run the thread on
+     * @return a chainable instance of Request, not guaranteed to be `this`
```
Comment (danielpassos) [inline core/src/main/java/org/aerogear/mobile/core/Request.java]: Markdown on the Javadoc?

File: core/src/main/java/org/aerogear/mobile/core/Request.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Requests may be asynchronous and need to be cancelled.
+     */
+    void cancel();
+
+    /**
+     * This tells the Request chain to cache its value and not to rerun any underlying generating
+     * function.
+     *
+     * @return a chainable instance of Request, not guaranteed to be `this`
```
Comment (danielpassos) [inline core/src/main/java/org/aerogear/mobile/core/Request.java]: Markdown on the Javadoc?

File: core/src/main/java/org/aerogear/mobile/core/Request.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+     * @return a request instance that represents the request the parameter was disconnected from.
+     */
+    Request<T> disconnect(Responder<T> responderToDisconnect);
+
+
+    /**
+     * Responses can be run off the calling thread. This method configures that.
+     *
+     * @param executorService service to Run the thread on
+     * @return a chainable instance of Request, not guaranteed to be `this`
```
Comment (danielpassos) [inline core/src/main/java/org/aerogear/mobile/core/Request.java]: Markdown on the Javadoc?

File: core/src/main/java/org/aerogear/mobile/core/reactive/AbstractRequest.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import android.support.annotation.NonNull;
+
+import org.aerogear.mobile.core.Request;
+import org.aerogear.mobile.core.Responder;
+import org.aerogear.mobile.core.utils.SanityCheck;
+
+/**
+ * This class performs wrapping and checking for subclasses.
+ */
+abstract class AbstractRequest<T> implements InternalRequest<T> {
```
Comment (danielpassos) [inline core/src/main/java/org/aerogear/mobile/core/reactive/AbstractRequest.java]: Should we merge `AbstractRequest` and `InternalRequest`?

File: core/src/main/java/org/aerogear/mobile/core/reactive/AbstractRequest.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Override
+    public final Request<T> respondOn(ExecutorService executorService) {
+        nonNull(executorService, "executorService");
+        return new RespondOnRequest<>(this, executorService);
+    }
+
+    @Override
+    public final Request<T> respondWith(@NonNull Responder<T> responder) {
+        SanityCheck.nonNull(responder, "responder");
```
Comment (danielpassos) [inline core/src/main/java/org/aerogear/mobile/core/reactive/AbstractRequest.java]: I see in some places you are using the static import and in other you are using the full path `SanityCheck.****`

File: core/src/main/java/org/aerogear/mobile/core/Request.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
+package org.aerogear.mobile.core;
```
Comment (danielpassos) [inline core/src/main/java/org/aerogear/mobile/core/Request.java]: Should we move this class into the reactive package since this is reactive related?

File: core/src/main/java/org/aerogear/mobile/core/Responder.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
+package org.aerogear.mobile.core;
```
Comment (danielpassos) [inline core/src/main/java/org/aerogear/mobile/core/Responder.java]: Should we move this class into the reactive package since this is reactive related?

**Comment by secondsun** [comment]
@danielpassos, @ziccardi  Applied your feedback

File: core/src/main/java/org/aerogear/mobile/core/reactive/CallableRequest.java (hunk 2 above)
Comment (wojta) [reply core/src/main/java/org/aerogear/mobile/core/reactive/CallableRequest.java]: You have there that sanity check but we should also have `@NonNull Callable<T> callable` there for it to work with IDE's checks and Kotlin's null-safety.
